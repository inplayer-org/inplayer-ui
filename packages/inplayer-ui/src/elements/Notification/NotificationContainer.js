import styled, { keyframes, css } from 'styled-components';
import { switchProp, prop } from 'styled-tools';
import colors from 'config/colors';
import { transparentize } from 'polished';

const notify = keyframes`
  0%,
  100% {
    opacity: 0;
    top: 50px;
  }

  5%,
  95% {
    opacity: 1;
    top: 80px;
  }
`;

const NotificationContainer = styled.div`
  box-sizing: border-box;
  border: 1px solid ${colors.gray};
  border-radius: 3px;
  background: ${colors.lightGray};
  color: ${colors.fontDarkGray};
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 1.125rem 2%;
  min-height: 10%;
  left: 15%;
  right: 15%;
  position: fixed;
  top: 80px;
  z-index: 120;
  box-shadow: 0 5px 20px rgba(0, 0, 0, 0.15);
  opacity: 0;
  animation: ${notify} ${prop('duration', 8)}s ease-in-out 0.2s 1;
  ${switchProp('variant', {
    success: css`
      border-color: ${colors.green};
      background: ${colors.paleGreen};

      strong {
        color: ${colors.green};
      }
    `,
    danger: css`
      border-color: ${colors.red};
      background: ${colors.paleRed};

      strong {
        color: ${colors.red};
      }
    `,
    warning: css`
      border-color: ${colors.yellow};
      background: ${colors.paleYellow};

      strong {
        color: ${colors.yellow};
      }
    `,
  })};

  i {
    margin-right: 0.75rem;
    position: relative;
    top: 2px;
  }

  p {
    margin: 0;
    max-width: none;
  }
`;

export const CloseIcon = styled.div`
  width: 18px;
  height: 18px;
  border-radius: 50%;
  align-self: flex-start;
  color: ${transparentize(0.4, colors.fontDarkGray)};
  cursor: pointer;
  transition: all 0.5s ease;

  &::before {
    content: 'âœ•';
    position: relative;
    left: 5px;
    font-size: 0.75rem;
    top: -1px;
  }

  &:hover {
    background: ${colors.white};
    color: ${colors.fontDarkGray};
  }
`;

export default NotificationContainer;
