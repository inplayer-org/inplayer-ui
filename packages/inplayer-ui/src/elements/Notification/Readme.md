The Notification exposes an API for creating notifications. The API consists of 4 functions: `Notification.create`, `Notification.success`, `Notification.danger` and `Notification.warning`. The latter 3 are simply abstractions over the first one which set the `variant` to `success`, `danger` and `warning` accordingly. If no variant is passed it will simply use a neutral grey color scheme (check the demo.).

The functions take 2 arguments: first one is an object which represents the props of the notification, the second one is optional and is a reference to a parent DOM element for the notification. If the parent DOM is not passed the notification will be mounted as a child on the `body` element. The way it is used is:

```jsx static
import { Notification, Button } from '@inplayer-org/inplayer-ui';

function Page(props) {
  return (
    <div>
      <Button
        onClick={() =>
          Notification.create({
            title: 'Notification',
            content: 'This is a regular notification',
            duration: 3,
          })
        }
      >
        Show notification
      </Button>
      <Button
        onClick={() =>
          Notification.success({
            title: 'Hooray!',
            content: 'This is a success notification',
            duration: 3,
          })
        }
      >
        Show success notification
      </Button>
      <Button
        onClick={() =>
          Notification.danger({
            title: 'Oh no!',
            content: 'This is a danger notification',
            duration: 3,
          })
        }
      >
        Show danger notification
      </Button>
      <Button
        onClick={() =>
          Notification.warning({
            title: 'Careful!',
            content: 'This is a warning notification',
            duration: 3,
          })
        }
      >
        Show warning notification
      </Button>
    </div>
  );
}
```

### Demo

```jsx
  <Button onClick={() => Notification.create({
    title: 'Notification',
    content: 'This is a regular notification',
    duration: 3,
  })}>Show notification</Button> <br />
  <Button onClick={() => Notification.success({
    title: 'Hooray!',
    content: 'This is a success notification',
    duration: 3,
  })}>Show success notification</Button> <br />
  <Button onClick={() => Notification.danger({
    title: 'Oh no!',
    content: 'This is a danger notification',
    duration: 3,
  })}>Show danger notification</Button> <br />
  <Button onClick={() => Notification.warning({
    title: 'Careful!',
    content: 'This is a warning notification',
    duration: 3,
  })}>Show warning notification</Button>
```
