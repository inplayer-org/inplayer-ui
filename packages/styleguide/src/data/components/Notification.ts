import { ComponentData } from "../types";

const usage = `
  import { Notification, Button } from '@inplayer-org/inplayer-ui';

  function Page(props) {
    return (
      <div>
        <Button
          onClick={() =>
            Notification.create({
              title: 'Notification',
              content: 'This is a regular notification',
              duration: 3,
            })
          }
        >
          Show notification
        </Button>
        <Button
          onClick={() =>
            Notification.success({
              title: 'Hooray!',
              content: 'This is a success notification',
              duration: 3,
            })
          }
        >
          Show success notification
        </Button>
        <Button
          onClick={() =>
            Notification.danger({
              title: 'Oh no!',
              content: 'This is a danger notification',
              duration: 3,
            })
          }
        >
          Show danger notification
        </Button>
        <Button
          onClick={() =>
            Notification.warning({
              title: 'Careful!',
              content: 'This is a warning notification',
              duration: 3,
            })
          }
        >
          Show warning notification
        </Button>
      </div>
    );
  }
`
const demoCode = `
<div>
  <Button onClick={() => Notification.create({
    title: 'Notification',
    content: 'This is a regular notification',
    duration: 3,
  })}>Show notification</Button> <br />
  <Button onClick={() => Notification.success({
    title: 'Hooray!',
    content: 'This is a success notification',
    duration: 3,
  })}>Show success notification</Button> <br />
  <Button onClick={() => Notification.danger({
    title: 'Oh no!',
    content: 'This is a danger notification',
    duration: 3,
  })}>Show danger notification</Button> <br />
  <Button onClick={() => Notification.warning({
    title: 'Careful!',
    content: 'This is a warning notification',
    duration: 3,
  })}>Show warning notification</Button>
</div>
`

const Notification: ComponentData = {
  description: `The Notification exposes an API for creating notifications. The API consists of 4 functions: Notification.create, Notification.success, Notification.danger and Notification.warning. The latter 3 are simply abstractions over the first one which set the variant to success, danger and warning accordingly. If no variant is passed it will simply use a neutral grey color scheme (check the demo.).

  The functions take 2 arguments: first one is an object which represents the props of the notification, the second one is optional and is a reference to a parent DOM element for the notification. If the parent DOM is not passed the notification will be mounted as a child on the body element. The way it is used is:`,
  path: `src/elements/Notification/index.js`,
  usage,
  examples: [
    {
      title: 'Demo',
      code: demoCode,
    }
  ],
  propsAndMethods: [
    { name: 'close', type: 'function', default: 'Required' },
    { name: 'content', type: 'string', default: 'Required' },
    { name: 'title', type: 'string', default: 'Required' },
    { name: 'className', type: 'string'},
    { name: 'duration', type: 'number', default: '{}' },
    { name: 'variant', type: 'enum', description: `One of: success, danger, warning`},
  ]
}

export default Notification;