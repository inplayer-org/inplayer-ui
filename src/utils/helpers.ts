import moment from 'moment';

export const getMonthOptions = (): Array<any> =>
  moment.months().map((label, value) => ({ value, displayName: label }));

export const getYearOptions = () => {
  const yearOptions: Array<any> = [];
  for (let i = moment().year() - 100; i <= moment().year(); i += 1) {
    yearOptions.push({ value: i.toString(), displayName: i.toString() });
  }

  return yearOptions;
};
