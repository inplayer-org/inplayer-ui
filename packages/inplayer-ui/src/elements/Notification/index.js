// @flow
import React, { useEffect } from 'react';
import ReactDOM from 'react-dom';

import NotificationContainer, { CloseIcon } from './NotificationContainer';

type NotificationVariant = 'success' | 'danger' | 'warning';

type NotificationProps = {
  title: string,
  content: string,
  variant?: NotificationVariant,
  duration?: number,
  className?: string,
  style?: Object,
  close: () => void,
};

const divs: Array<HTMLDivElement> = [];

function useTimeout(callback, duration) {
  useEffect(() => {
    if (typeof duration === 'number' && duration >= 0) {
      const id = setTimeout(callback, duration);
      return () => clearTimeout(id);
    }
  }, [duration]);
}

const Notification = ({
  title,
  content,
  variant,
  duration,
  className,
  close,
  style,
}: NotificationProps) => {
  useTimeout(close, duration * 1000);
  return (
    <NotificationContainer
      variant={variant}
      duration={duration}
      className={className}
      style={style}
    >
      <div>
        <strong>{title}</strong> {content}
      </div>
      <CloseIcon onClick={close} />
    </NotificationContainer>
  );
};

function destroy(index) {
  if (divs[index]) {
    ReactDOM.unmountComponentAtNode(divs[index]);
    document.body.removeChild(divs[index]);
  }
}

function create(props: NotificationProps, parentDiv: ?HTMLDivElement) {
  const divIndex = divs.length;
  const notificationInstance = <Notification {...props} close={() => destroy(divIndex)} />;
  const div = document.createElement('div');
  if (parentDiv) {
    parentDiv.appendChild(div);
  } else if (document.body) {
    document.body.appendChild(div);
  }

  ReactDOM.render(notificationInstance, div);
  divs.push(div);
}

Notification.create = create;

Notification.success = function success(props: NotificationProps, parentDiv: ?HTMLDivElement) {
  create({ ...props, variant: 'success' }, parentDiv);
};

Notification.warning = function warning(props: NotificationProps, parentDiv: ?HTMLDivElement) {
  create({ ...props, variant: 'warning' }, parentDiv);
};

Notification.danger = function danger(props: NotificationProps, parentDiv: ?HTMLDivElement) {
  create({ ...props, variant: 'danger' }, parentDiv);
};

Notification.defaultProps = {
  variant: '',
  duration: 0,
  className: '',
  style: {},
};

/** @component */
export default Notification;
