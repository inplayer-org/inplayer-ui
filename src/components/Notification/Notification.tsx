import React, { useEffect } from 'react';
import ReactDOM from 'react-dom';
import NotificationContainer, { CloseIcon } from './NotificationContainer';

export type NotificationVariant = 'success' | 'danger' | 'warning' | '';

type Props = {
  /**
   * Notification title
   */
  title: string;
  /**
   * Notification content
   */
  content: string;
  /**
   * Variant type
   */
  variant?: NotificationVariant;
  /**
   * Duration (in seconds) of the notification before it disappears
   */
  duration?: number;
  /**
   * Function to close the notification
   */
  close?: () => void;
};

const divs: Array<HTMLDivElement> = [];

const useTimeout = (callback: () => void, duration: number) => {
  useEffect(() => {
    if (typeof duration === 'number' && duration >= 0) {
      const id = setTimeout(callback, duration);
      return () => clearTimeout(id);
    }
  }, [duration]);
};

const destroy = (index: number) => {
  if (divs[index]) {
    ReactDOM.unmountComponentAtNode(divs[index]);
    document.body.removeChild(divs[index]);
  }
};

const Notification = ({
  title,
  content,
  variant = '',
  duration = 0,
  close = () => destroy(divs.length),
}: Props) => {
  useTimeout(close, duration * 1000);
  return (
    <NotificationContainer variant={variant} duration={duration}>
      <div>
        <strong>{title}</strong> {content}
      </div>
      <CloseIcon onClick={close} />
    </NotificationContainer>
  );
};

const create = (props: Props, parentDiv?: HTMLDivElement) => {
  const divIndex = divs.length;
  const notificationInstance = <Notification {...props} close={() => destroy(divIndex)} />;
  const div = document.createElement('div');
  if (parentDiv) {
    parentDiv.appendChild(div);
  } else if (document.body) {
    document.body.appendChild(div);
  }

  ReactDOM.render(notificationInstance, div);
  divs.push(div);
};

const success = (props: Props, parentDiv?: HTMLDivElement) => {
  create({ ...props, variant: 'success' }, parentDiv);
};

const warning = (props: Props, parentDiv?: HTMLDivElement) => {
  create({ ...props, variant: 'warning' }, parentDiv);
};

const danger = (props: Props, parentDiv?: HTMLDivElement) => {
  create({ ...props, variant: 'danger' }, parentDiv);
};

Notification.create = create;
Notification.success = success;
Notification.warning = warning;
Notification.danger = danger;

export default Notification;
