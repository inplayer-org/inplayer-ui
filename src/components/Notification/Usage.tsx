/* eslint-disable react/button-has-type */
import React from 'react';
import Notification from '.';

const Usage = () => (
  <>
    <button
      onClick={() =>
        Notification.create({
          title: 'Notification',
          content: 'This is a regular notification',
          duration: 3,
        })
      }
    >
      REGULAR NOTIFICATION
    </button>
    <button
      onClick={() =>
        Notification.create({
          title: 'Hooray! ',
          variant: 'success',
          content: 'This is a success notification',
          duration: 3,
        })
      }
    >
      SUCCESS NOTIFICATION
    </button>
    <button
      onClick={() =>
        Notification.create({
          title: 'Warning! ',
          variant: 'warning',
          content: 'This is a warning notification',
          duration: 3,
        })
      }
    >
      WARNING NOTIFICATION
    </button>
    <button
      onClick={() =>
        Notification.create({
          title: 'Danger! ',
          variant: 'danger',
          content: 'This is a danger notification',
          duration: 3,
        })
      }
    >
      DANGER NOTIFICATION
    </button>
  </>
);

export default Usage;
