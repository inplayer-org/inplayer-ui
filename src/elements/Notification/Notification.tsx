import React, { useEffect } from 'react';
import ReactDOM from 'react-dom';

import NotificationContainer, { CloseIcon } from './NotificationContainer';

type NotificationVariant = 'success' | 'danger' | 'warning';

interface NotificationProps {
  title: string;
  content: string;
  duration: number;
  variant?: NotificationVariant;
  className?: string;
  style?: Record<string, any>;
  close?: () => void;
}

const divs: Array<HTMLDivElement> = [];

function useTimeout(callback: any, duration: any) {
  useEffect(() => {
    if (duration >= 0) {
      const id = setTimeout(callback, duration);
      return () => clearTimeout(id);
    }
  }, [duration]);
}

const Notification = ({
  title,
  content,
  duration,
  variant,
  className,
  close,
  style,
}: NotificationProps) => {
  useTimeout(close, duration * 1000);
  return (
    <NotificationContainer
      variant={variant}
      duration={duration}
      className={className}
      style={style}
    >
      <div>
        <strong>{title} </strong> {content}
      </div>
      <CloseIcon onClick={close} />
    </NotificationContainer>
  );
};

function destroy(index: any) {
  if (divs[index]) {
    ReactDOM.unmountComponentAtNode(divs[index]);
    document.body.removeChild(divs[index]);
  }
}

function create(props: NotificationProps, parentDiv?: HTMLDivElement) {
  const divIndex = divs.length;
  const notificationInstance = <Notification {...props} close={() => destroy(divIndex)} />;
  const div = document.createElement('div');
  if (parentDiv) {
    parentDiv.appendChild(div);
  } else if (document.body) {
    document.body.appendChild(div);
  }

  ReactDOM.render(notificationInstance, div);
  divs.push(div);
}

Notification.create = create;

Notification.success = function success(props: NotificationProps, parentDiv?: HTMLDivElement) {
  create({ ...props, variant: 'success' }, parentDiv);
};

Notification.warning = function warning(props: NotificationProps, parentDiv?: HTMLDivElement) {
  create({ ...props, variant: 'warning' }, parentDiv);
};

Notification.danger = function danger(props: NotificationProps, parentDiv?: HTMLDivElement) {
  create({ ...props, variant: 'danger' }, parentDiv);
};

Notification.defaultProps = {
  variant: '',
  duration: 0,
  className: '',
  style: {},
};

export default Notification;
