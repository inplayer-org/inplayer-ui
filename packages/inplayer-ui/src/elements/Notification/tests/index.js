import { generateSnapshots } from 'packages/inplayer-ui/src/elements/Notification/tests/helpers';

import Notification from '../index';

generateSnapshots(Notification);
