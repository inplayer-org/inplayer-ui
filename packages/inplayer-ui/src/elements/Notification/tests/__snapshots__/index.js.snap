// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification renders correctly 1`] = `
.c0 {
  box-sizing: border-box;
  border: 1px solid #d7dde5;
  border-radius: 3px;
  background: #F6F8FB;
  color: #3e4348;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 1.125rem 2%;
  min-height: 10%;
  left: 15%;
  right: 15%;
  position: fixed;
  top: 80px;
  z-index: 120;
  box-shadow: 0 5px 20px rgba(0,0,0,0.15);
  opacity: 0;
  -webkit-animation: njGPL 0s ease-in-out 0.2s 1;
  animation: njGPL 0s ease-in-out 0.2s 1;
}

.c0 i {
  margin-right: 0.75rem;
  position: relative;
  top: 2px;
}

.c0 p {
  margin: 0;
  max-width: none;
}

.c1 {
  width: 18px;
  height: 18px;
  border-radius: 50%;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  color: rgba(62,67,72,0.6);
  cursor: pointer;
  -webkit-transition: all 0.5s ease;
  transition: all 0.5s ease;
}

.c1::before {
  content: 'âœ•';
  position: relative;
  left: 5px;
  font-size: 0.75rem;
  top: -1px;
}

.c1:hover {
  background: #fff;
  color: #3e4348;
}

<div
  className="c0"
  style={Object {}}
>
  <div>
    <strong />
     
  </div>
  <div
    className="c1"
  />
</div>
`;
